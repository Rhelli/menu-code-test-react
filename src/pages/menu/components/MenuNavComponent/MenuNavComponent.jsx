import React from 'react';
import { v4 as uuidv4 } from 'uuid';
import styles from './MenuNavComponent.module.scss';

const MenuNavComponent = ({ orders, setOrderStatus }) => {
  const names = Object.keys(orders);

  return (
    <header className={styles.menuNavContainer}>
      <div className={styles.menuTitle}>
        <h1>Menu</h1>
      </div>
      <div className={styles.menuNavButtons}>
        {
          names.map((name, i) => (
            <button
              onClick={() => setOrderStatus(name)}
              id={i}
              style={orders[name].color}
              key={uuidv4()}
            >
              {name}
            </button>
          ))
        }
        <button onClick={() => setOrderStatus('Sharing')} className={styles.sharedButton}>
          Sharing
        </button>
      </div>
    </header>
  );
};

export default MenuNavComponent;
