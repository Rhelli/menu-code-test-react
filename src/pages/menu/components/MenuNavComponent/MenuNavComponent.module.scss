.menuNavContainer {
  display: flex;
  flex-flow: row nowrap;
  width: 100%;
  height: auto;
  padding: 20px 15px;
  border: 2px solid #707070;
  border-radius: 3px;

  .menuNavButtons {
    display: flex;
    width: auto;
    margin-left: auto;

    button {
      border: none;
      outline: none;
      color: #154360;
      padding: 6px 12px;
      font-size: 12px;
      margin-left: 5px;
      border-radius: 3px;
      transition: 0.1s;
      opacity: 0.5;

      &:hover {
        opacity: 0.75;
        transition: 0.1s;
      }

      &:active,
      &:focus {
        opacity: 1;
      }
    }

    .sharedButton {
      background-color: #37DACE;
    }
  }
}


