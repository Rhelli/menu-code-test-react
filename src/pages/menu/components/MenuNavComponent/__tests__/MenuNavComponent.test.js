import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MenuNavComponent from '../MenuNavComponent';

Enzyme.configure({ adapter: new Adapter() });

const orders = {
  Sam: {
    color: {
      backgroundColor: '#d3445d'
    },
    starters: {},
    mains: {},
    desserts: {}
  },
  Frodo: {
    color: {
      backgroundColor: '#404040'
    },
    starters: {},
    mains: {},
    desserts: {}
  }
};
const setCurrentGuest = jest.fn();
const currentGuest = 'Sam';

const menuNavComponent = (
  <MenuNavComponent
  orders={orders}
  setCurrentGuest={setCurrentGuest}
  currentGuest={currentGuest}
  />
);

describe('The appearance and functionality of the MenuNavComponent', () => {
  const container = mount(
    <MenuNavComponent
    orders={orders}
    setCurrentGuest={setCurrentGuest}
    currentGuest={currentGuest}
    />
  );
  it('Matches the previous snapshot taken', () => {
    const tree = renderer.create(menuNavComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders Correctly', () => {
    mount(<MenuNavComponent orders={orders} setCurrentGuest={setCurrentGuest} currentGuest={currentGuest} />);
  });

  it('Contains the master container with the correct attributes', () => {
    expect(container.find('.menuNavContainer').type()).toBe('header');
  });

  it('Contains buttons for each user provided from the orders prop', () => {
    expect(container.find('.menuNavButtons').children()).toHaveLength(2);
    expect(container.find('.menuNavButtons').childAt(0).text()).toBe('Sam');
    expect(container.find('.menuNavButtons').childAt(1).text()).toBe('Frodo');
  });

  it('Automatically focuses the first guests button on startup and unfocuses the other', () => {
    expect(container.find('.menuNavButtons').childAt(0).hasClass('selected')).toBe(true);
    expect(container.find('.menuNavButtons').childAt(1).hasClass('unselected')).toBe(true);
  });

  it('Should update the currentGuest state when an unfocused button is clicked', () => {
    const handleClick = jest.spyOn(React, 'useState');
    handleClick.mockImplementation((currentGuest) => [currentGuest, setCurrentGuest]);
    container.find('.menuNavButtons').childAt(1).simulate('click');
    expect(currentGuest).toBe('Sam');
  });

  it('Paints the button the color assigned to the guest on welcome', () => {
    expect(container.find('.menuNavButtons').childAt(0).prop('style')).toEqual({ backgroundColor: '#d3445d' });
  });
});
