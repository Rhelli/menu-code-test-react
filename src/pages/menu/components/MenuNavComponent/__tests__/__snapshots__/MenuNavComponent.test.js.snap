// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The appearance and functionality of the MenuNavComponent Matches the previous snapshot taken 1`] = `
<header
  className="menuNavContainer"
>
  <div
    className="menuTitle"
  >
    <h1>
      Menu
    </h1>
  </div>
  <div
    className="menuNavButtons"
  >
    <button
      className="selected"
      id={0}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#d3445d",
        }
      }
    >
      Sam
    </button>
    <button
      className="unselected"
      id={1}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#404040",
        }
      }
    >
      Frodo
    </button>
  </div>
</header>
`;
